### DISCUSSION TITLE
Support for constant propagation

### DISCUSSION THREAD

**Discussion by llooFlashooll** [Fraunhofer-AISEC/cpg#1831] [Q&A] [answered]
Hi folks, I really appreciate your work.

However, the documentation of this repo says it supports constant propagation, but I cannot find the concrete code implementation or the corresponding test cases. Can you give me further instructions? Thank you!

**Comment by oxisto** [comment, marked as answer]
You need to include the `cpg-analysis` module. You can then use the `ValueEvaluator` class to evaluate simple constants such as strings and numbers in most languages. You can also override the default behaviour if you specify your own implementation of `notDefined`. There is also a shortcut you can use: `evaluate()`, so if you have a node that you suspect will contain a constant you can do

```Kotlin
import de.fraunhofer.aisec.cpg.graph.*

val node = /* ... */
val value = node.evaluate()
```

This is equivalent to `ValueEvaluator().evaluate(node)`

You can find some more test cases here: https://github.com/Fraunhofer-AISEC/cpg/blob/main/cpg-analysis/src/test/kotlin/de/fraunhofer/aisec/cpg/analysis/ValueEvaluatorTest.kt

**Reply by llooFlashooll** [reply]
Thank you very much for your detailed reply! It seems that the constant propagation algorithm is simple. Let me try it out and see the results!

**Reply by oxisto** [reply]
> Thank you very much for your detailed reply! It seems that the constant propagation algorithm is simple. Let me try it out and see the results!

Yes it is very simple at the moment. We are currently working on improving this with some ideas in the direction of abstract interpretation. https://github.com/Fraunhofer-AISEC/cpg/pull/1759
